File: examples/CBuilder/Compound/Locking/LockManagerAdminClient/dmLockManagerAdmin.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "dmLockManagerAdmin.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BoldAbstractLockManagerAdminHandle"
#pragma link "BoldClientHandles"
#pragma link "BoldComClientHandles"
#pragma link "BoldHandle"
#pragma link "BoldLockManagerAdminHandleCom"
#pragma link "BoldSubscription"
#pragma link "BoldUtils"
#pragma resource "*.dfm"
TdmMain *dmMain;
//---------------------------------------------------------------------------
__fastcall TdmMain::TdmMain(TComponent* Owner)
  : TDataModule(Owner)
{
}
//---------------------------------------------------------------------------
void FreeObjects(const TStringList* List)
{
  Integer i;
  TObject* obj;

  if (!List)
    exit(i);

  for (i = 0; i < List->Count; i++)
  {
    obj = List->Objects[i];
    obj->Free(); // FreeAndNil(obj);
  }
}
//---------------------------------------------------------------------------
void AssignObjects(const TStringList* Source, TStringList* Destination )
{
  if ( (Source) && (Destination) )
    for (int i = 0; i < Source->Count; i++)
      Destination->Objects[i] = Source->Objects[i];
}
//---------------------------------------------------------------------------
/* TClientInfo */
__fastcall TClientInfo::TClientInfo(AnsiString IdString)
  : TObject()
{
  fIdString = IdString;
  fLocks = new TStringList();
}

__fastcall TClientInfo::~TClientInfo()
{
  FreeObjects(fLocks);
  // FreeAndNil(fLocks);
}
//---------------------------------------------------------------------------
/* TLockInfo */
__fastcall TLockInfo::TLockInfo(AnsiString Duration, AnsiString LockName)
  : TObject()
{
  fDuration = Duration;
  fLockName = LockName;
  fClients = new TStringList();
}

__fastcall TLockInfo::~TLockInfo()
{
  FreeObjects(fClients);
  // FreeAndNil(fClients);
}
//---------------------------------------------------------------------------
void __fastcall TdmMain::GetClients()
{
  TStringList *lClients, *lLocks, *lLockDurations;
  OleVariant vClients, vLocks, vLockDurations;
  TClientInfo* clientInfo;
  TLockInfo* lockInfo;
  Integer i, j, idx;
  long l;

  if (BoldComConnectionHandle1->Connected)
  {
    lClients = new TStringList();
    lLocks = new TStringList();
    lLockDurations = new TStringList();
    try
    {
      if (!fClients)
        fClients = new TStringList();
      if (!fLocks)
        fLocks = new TStringList();
      if (!BoldLockManagerAdminHandleCom1->Active)
        BoldLockManagerAdminHandleCom1->Active = true;

      // listallclients         (1)
      if (ViewAll)
        BoldLockManagerAdminHandleCom1->LockManagerAdmin->ListAllClients(vClients, l);
      else
        BoldLockManagerAdminHandleCom1->LockManagerAdmin->ListLockingClients(vClients, l);

      BoldVariantToStrings(vClients, lClients);

      // get locks for clients
      BoldLockManagerAdminHandleCom1->LockManagerAdmin->LocksForClients(vClients, vLocks, vLockDurations, l);
      BoldVariantToStrings(vLocks, lLocks);
      BoldVariantToStrings(vLockDurations, lLockDurations);

      if (lLocks)
        for (i = 0; i < lLocks->Count; i++)
        {
          lockInfo = new TLockInfo(lLockDurations->Strings[i], lLocks->Values[lLocks->Names[i]]);
          lLocks->Strings[i] = lLocks->Names[i];
          lLocks->Objects[i] = lockInfo;
        }

      if (lClients)
        for (i = 0; i < lClients->Count; i++)
        {
          clientInfo = new TClientInfo(lClients->Values[lClients->Names[i]]);
          lClients->Strings[i] = lClients->Names[i];
          lClients->Objects[i] = clientInfo;

          j = lLocks->IndexOf(lClients->Strings[i]);
          while (j != -1)
          {
            idx = clientInfo->Locks->Add(lLocks->Strings[j]);
            clientInfo->Locks->Strings[j] = ((TLockInfo*)(lLocks->Objects[j]))->LockName;
            clientInfo->Locks->Objects[idx] = lLocks->Objects[j];
            lLocks->Delete(j);
            j = lLocks->IndexOf(lClients->Strings[i]);
          }
        }

      Clients = lClients;
    }
    __finally
    {
      lClients->Free(); // FreeAndNil(lClients);
      lLocks->Free(); // FreeAndNil(lLocks);
      lLockDurations->Free(); // FreeAndNil(lLockDurations);
    }
  }
}
//---------------------------------------------------------------------------
Boolean __fastcall TdmMain::GetLockManagerSuspended()
{
  Boolean result = false;
  if (BoldComConnectionHandle1->Connected)
  {
    if (!BoldLockManagerAdminHandleCom1->Active)
      BoldLockManagerAdminHandleCom1->Active = true;
    result = BoldLockManagerAdminHandleCom1->LockManagerAdmin->LockManagerSuspended;
  }
  return result;
}
//---------------------------------------------------------------------------
AnsiString __fastcall TdmMain::GetServerName()
{
  return Trim(frmMain->edServerName->Text);
}
//---------------------------------------------------------------------------
void __fastcall TdmMain::KillClient(Integer ClientId)
{
  try
  {
    if (BoldComConnectionHandle1->Connected)
    {
      if (!BoldLockManagerAdminHandleCom1->Active)
        BoldLockManagerAdminHandleCom1->Active = true;
      long l;
      BoldLockManagerAdminHandleCom1->LockManagerAdmin->KillClient(ClientId, l);
      dmMain->GetClients();
    }
  } catch (Exception &e)
  {
    throw e;
  }
}
//----------------------------------------------------------------------------
void __fastcall TdmMain::SetClients(const TStringList* Value)
{
  Integer i, idx;

  FreeObjects(fClients);
  fClients->Clear();
  fClients->Assign((TPersistent*)Value);
  AssignObjects(Value, fClients);
  if ( (frmMain) && (frmMain->PageControl1->ActivePage->Name == "tsClients") )
  {
    frmMain->lbClients->Clear();
    for (i = 0; i < fClients->Count; i++)
    {
      //TVarRec v[] = { fClients->Strings[i], ((TClientInfo*)(fClients->Objects[i]))->IdString };
      idx = frmMain->lbClients->Items->Add( fClients->Strings[i] + "=" + ((TClientInfo*)(fClients->Objects[i]))->IdString );// Format("%s=%s", v, ARRAYSIZE(v) - 1));
      frmMain->lbClients->Items->Objects[idx] = fClients->Objects[i];
    }
  }
}
//----------------------------------------------------------------------------
void __fastcall TdmMain::SetLockManagerSuspended(const Boolean Value)
{
  if (BoldComConnectionHandle1->Connected)
  {
    if (!BoldLockManagerAdminHandleCom1->Active)
      BoldLockManagerAdminHandleCom1->Active = true;
    BoldLockManagerAdminHandleCom1->LockManagerAdmin->LockManagerSuspended = Value;
  }
}
//----------------------------------------------------------------------------
void __fastcall TdmMain::SetLocks(const TStringList* Value)
{
  FreeObjects(fLocks);
  fLocks->Clear();
  fLocks->Assign((TPersistent*)Value);
  AssignObjects(Value, fLocks);
}
//----------------------------------------------------------------------------
void __fastcall TdmMain::SetViewAll(const Boolean Value)
{
  if (Value != fViewAll)
  {
    fViewAll = Value;
    GetClients();
  }
}

File: examples/Delphi/Compound/Locking/LockManagerAdminClient/dmLockManagerAdmin.pas
unit dmLockManagerAdmin;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  BoldAbstractLockManagerAdminHandle, BoldLockManagerAdminHandleCom,
  BoldSubscription, BoldHandle, BoldClientHandles, BoldComClientHandles;

type

  TClientInfo = class
  private
    fIdString: string;
    fLocks: TStringList;
  public
    constructor Create(IdString: string);
    destructor Destroy; override;
    property IdString: string read fIdString write fIdString;
    property Locks: TStringList read fLocks write fLocks;
  end;

  TLockInfo = class
  private
    fDuration: string;
    fLockName: string;
    fClients: TStringList;
  public
    constructor Create(Duration: string; LockName:string);
    destructor Destroy; override;
    property Duration: string read fDuration write fDuration;
    property LockName: string read fLockName write fLockName;
  end;

  TdmMain = class(TDataModule)
    BoldComConnectionHandle1: TBoldComConnectionHandle;
    BoldLockManagerAdminHandleCom1: TBoldLockManagerAdminHandleCom;
  private
    { Private declarations }
    fClients: TStringList;
    fLocks: TStringList;
    fViewAll: Boolean;
    procedure SetLocks(const Value: TStringList);
    procedure SetClients(const Value: TStringList);
    procedure SetViewAll(const Value: Boolean);
    function GetLockManagerSuspended: Boolean;
    procedure SetLockManagerSuspended(const Value: Boolean);
    function getServerName: string;
  public
    { Public declarations }
    procedure GetClients;
    procedure KillClient(ClientId: integer);
    property LockManagerSuspended: Boolean read GetLockManagerSuspended write SetLockManagerSuspended;
    property Clients: TStringList read fClients write SetClients;
    property Locks: TStringList read fLocks write SetLocks;
    property ViewAll: Boolean read fViewAll write SetViewAll;
    property ServerName: string read getServerName;
  end;

  procedure FreeObjects(const List: TStringList);
  procedure AssignObjects(const Source, Destination: TStringList);

var
  dmMain: TdmMain;


implementation

uses
  BoldUtils,
  MainForm;
{$R *.DFM}

procedure FreeObjects(const List: TStringList);
var
  i: integer;
  obj : TObject;
begin
  if not Assigned(List) then
    Exit;
  for i:= 0 to list.Count - 1 do
  begin
    obj := list.Objects[i];
    FreeAndNil(obj);
  end;
end;

procedure AssignObjects(const Source, Destination: TStringList );
var
  i: integer;
begin
  if not Assigned(Source) or not Assigned(Destination) then
    Exit;
  for i:= 0 to source.Count - 1 do
  begin
    Destination.Objects[i] := Source.Objects[i];
  end;
end;

{ TClientInfo }

constructor TClientInfo.Create(IdString: string);
begin
  inherited Create;
  fIdString := IdString;
  fLocks := TStringList.Create;
end;

destructor TClientInfo.Destroy;
begin
  FreeObjects(fLocks);
  FreeAndNil(fLocks);
  inherited;
end;

{ TLockInfo }

constructor TLockInfo.Create(Duration: string; LockName: string);
begin
  inherited Create;
  fDuration:= Duration;
  fLockName := LockName;
  fClients := TStringList.Create;
end;

destructor TLockInfo.Destroy;
begin
  FreeObjects(fClients);
  FreeAndNil(fClients);
  inherited;
end;

{ TdmMain }

procedure TdmMain.GetClients;
var
  lClients, lLocks, lLockDurations: TStringList;
  vClients, vLocks, vlockDurations: OleVariant;
  clientInfo: TClientInfo;
  lockInfo: TLockInfo;
  i, j, idx: integer;
begin
  if BoldComConnectionHandle1.Connected then
  begin
    lClients := TStringList.Create;
    lLocks := TStringList.Create;
    lLockDurations := TStringList.Create;
    try
      if not Assigned(fClients) then
        fClients := TStringList.Create;
      if not Assigned(fLocks) then
        fLocks := TStringList.Create;
      if not BoldLockManagerAdminHandleCom1.Active then
        BoldLockManagerAdminHandleCom1.Active := True;

      // listallclients         (1)
      if ViewAll then
        BoldLockManagerAdminHandleCom1.LockManagerAdmin.ListAllClients(vClients)
      else
        BoldLockManagerAdminHandleCom1.LockManagerAdmin.ListLockingClients(vclients);
      BoldVariantToStrings(vClients, lClients);

      // get locks for clients
      BoldLockManagerAdminHandleCom1.LockManagerAdmin.LocksForClients(vclients, vLocks, vLockDurations);
      BoldVariantToStrings(vLocks, lLocks);
      BoldVariantToStrings(vLockDurations, lLockDurations);
      for i := 0 to lLocks.Count - 1 do
      begin
        lockInfo := TLockInfo.Create(lLockDurations[i], lLocks.Values[lLocks.Names[i]]);
        lLocks[i] := lLocks.Names[i];
        lLocks.Objects[i] := lockinfo;
      end;

      for i:= 0 to (lClients.Count - 1) do
      begin
        clientInfo := TClientInfo.Create(lClients.Values[lClients.Names[i]]);
        lClients[i] := lClients.Names[i];
        lClients.Objects[i]:= clientinfo;
        j := lLocks.IndexOf(lClients[i]);
        while j <> -1 do
        begin
          idx := clientinfo.Locks.Add(lLocks[j]);
          clientinfo.Locks[j] := (lLocks.Objects[j] as TLockInfo).LockName;
          clientinfo.Locks.Objects[idx] := lLocks.Objects[j];
          lLocks.Delete(j);
          j := lLocks.IndexOf(lClients[i]);
        end
      end;

      Clients := lClients;
    finally
      FreeAndNil(lClients);
      FreeAndNil(lLocks);
      FreeAndNil(lLockDurations);
    end;
  end;
end;

function TdmMain.GetLockManagerSuspended: Boolean;
begin
  if BoldComConnectionHandle1.Connected then
  begin
    if not BoldLockManagerAdminHandleCom1.Active then
      BoldLockManagerAdminHandleCom1.Active := True;
    Result := BoldLockManagerAdminHandleCom1.LockManagerAdmin.LockManagerSuspended;
  end
  else
    Result := false;
end;

function TdmMain.getServerName: string;
begin
  Result := trim(fMain.edServerName.Text);
end;

procedure TdmMain.KillClient(ClientId: integer);
begin
  try
    if BoldComConnectionHandle1.Connected then
    begin
      if not BoldLockManagerAdminHandleCom1.Active then
        BoldLockManagerAdminHandleCom1.Active := True;
      BoldLockManagerAdminHandleCom1.LockManagerAdmin.KillClient(ClientId);
      dmMain.GetClients;
    end;
  finally
  end;
end;

procedure TdmMain.SetClients(const Value: TStringList);
var
  i, idx: integer;
begin
  FreeObjects(fClients);
  fClients.Clear;
  fClients.Assign(Value);
  AssignObjects(Value, fClients);
  if Assigned(MainForm.fMain) and (MainForm.FMain.PageControl1.ActivePage.Name = 'tsClients') then
  begin
    MainForm.FMain.lbclients.Clear;
    for i:= 0 to fClients.Count - 1 do
    begin
      idx := MainForm.FMain.lbclients.Items.Add(Format('%s=%s', [fClients[i], (fClients.Objects[i] as TClientInfo).IdString]));
      MainForm.FMain.lbclients.Items.Objects[idx] := fClients.Objects[i];
    end;
  end;
end;

procedure TdmMain.SetLockManagerSuspended(const Value: Boolean);
begin
  if BoldComConnectionHandle1.Connected then
  begin
    if not BoldLockManagerAdminHandleCom1.Active then
      BoldLockManagerAdminHandleCom1.Active := True;
    BoldLockManagerAdminHandleCom1.LockManagerAdmin.LockManagerSuspended := Value;
  end;
end;

procedure TdmMain.SetLocks(const Value: TStringList);
begin
  FreeObjects(fLocks);
  fLocks.Clear;
  fLocks.Assign(Value);
  AssignObjects(Value, fLocks);
end;

procedure TdmMain.SetViewAll(const Value: Boolean);
begin
  if (Value <> fViewAll) then
  begin
    fViewAll := Value;
    getClients;
  end;
end;

end.
